stablehlo/dialect/Base.cpp

```diff
@@ -146,5 +146,24 @@ LogicalResult verifyBounds(ArrayRef<int64_t> bounds, ShapedType type,
   return success();
 }
 
+ArrayRef<int64_t> encodingToBounds(Attribute encoding) {
+  if (!encoding) return {};
+  if (auto boundedAttr = encoding.dyn_cast_or_null<BoundedAttrInterface>())
+    return boundedAttr.getBounds();
+  return {};
+}
+
+FailureOr<Attribute> boundsToEncoding(Attribute prototype,
+                                      ArrayRef<int64_t> bounds,
+                                      Optional<Location> loc) {
+  if (bounds.empty()) return prototype;
+  if (!prototype)
+    return emitOptionalError(loc,
+                             "Expect an prototype attribute to obtain the "
+                             "underlying dialect but got none");
```

Comment:
Is it possible that there will be an expected situation where we have non-empty bounds and empty prototype? I don't think so, so I believe that this should be `report_fatal_error` and the return type will no longer need `FailureOr`. The function will also no longer need a `Location`.

Reply:
Yeah this makes sense: report_fatal_error used in a totally unexpected case which should not happen at all. Fixed.